### PR TITLE
Fixes issue where Mint returns error "not enough valid signatures" in wrong scenario

### PR DESCRIPTION
**Pull request by areebusmani**
## Fixes
1. Return appropriate error when there is an error in finding number of authorisers.
2. Return from Mint function once there is error returned from verifySignatures function

## Changes
1. Add unit test case for covering the scenario where number of authorisers are 0 but signatures are non empty.

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/zcnsc/mint_test.go [hunk 1]
```diff
@@ -81,6 +81,22 @@ func Test_EmptySignaturesShouldFail(t *testing.T) {
 	require.Error(t, err)
 }
 
+func Test_EmptyAuthorizersShouldFail(t *testing.T) {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/zcnsc/mint_test.go]: add test where number of signatures are not empty

**Review by dabasov — DISMISSED** [review]
lgtm

File: code/go/0chain.net/smartcontract/zcnsc/mint_test.go (hunk 1 above)
Comment (areebusmani) [reply code/go/0chain.net/smartcontract/zcnsc/mint_test.go]: Done

File: code/go/0chain.net/smartcontract/zcnsc/mint.go [hunk 2]
```diff
@@ -49,6 +49,12 @@ func (zcn *ZCNSmartContract) Mint(trans *transaction.Transaction, inputData []by
 	}
 
 	numAuth, err := getAuthorizerCount(ctx)
+	if err != nil {
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/zcnsc/mint.go:51]: The err is nil even the numAuth does not exist. So we should also check if numAuth is 0.

File: code/go/0chain.net/smartcontract/zcnsc/mint_test.go [hunk 3]
```diff
@@ -81,6 +81,45 @@ func Test_EmptySignaturesShouldFail(t *testing.T) {
...
+	ctx.authorizers = nil
+
+	contract := CreateZCNSmartContract()
+	payload, err := CreateMintPayload(ctx, defaultClient)
+	require.NoError(t, err)
+
+	transaction, err := CreateTransaction(defaultClient, "mint", payload.Encode(), ctx)
+	require.NoError(t, err)
+
+	_, err = contract.Mint(transaction, payload.Encode(), ctx)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/zcnsc/mint_test.go:153]: The err is payload decoding error, please check the err message. If this is an intended test,  then we need to rename the Test to `Test_EmptySignaturehouldFail`.

File: code/go/0chain.net/smartcontract/zcnsc/mint_test.go [hunk 4]
```diff
@@ -81,6 +81,45 @@ func Test_EmptySignaturesShouldFail(t *testing.T) {
...
+	var signatures []*AuthorizerSignature
+	for i := 0; i < 10; i++ {
+		signatures = append(signatures, &AuthorizerSignature{})
+	}
+	payload.Signatures = signatures
+
+	transaction, err := CreateTransaction(defaultClient, "mint", payload.Encode(), ctx)
+	require.NoError(t, err)
+
+	_, err = contract.Mint(transaction, payload.Encode(), ctx)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/zcnsc/mint_test.go:175]: The `numAuth` is not empty in the `Mint` contract, the error is returned when verifying signatures, please check the place where the error is returned and make sure the tests do test the code we want.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/zcnsc/mint.go (hunk 2 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/zcnsc/mint.go:51]: I don't think we should check it, since we should invalidate any number of signatures in case we have 0 authorizers. In general, it should not differ how much authorizers we do ave 0, 1 or k, it should work the same way

Comment (peterlimg) [reply code/go/0chain.net/smartcontract/zcnsc/mint.go:51]: > since we should invalidate any number of signatures. in case we have 0 authorizers. 

Yes, that's what I mean, `numAuth` is the number of authorizers, and we should stop the Mint if it is 0, otherwise the zero division panic would happen as we have encountered.

File: code/go/0chain.net/smartcontract/zcnsc/mint_test.go (hunk 3 above)
Comment (areebusmani) [reply code/go/0chain.net/smartcontract/zcnsc/mint_test.go:153]: Removed this test cases as the scenario where number of authorizers are 0 is already covered in another test (Test_EmptyAuthorizersNonemptySignaturesShouldFail) that I have added in this PR.

File: code/go/0chain.net/smartcontract/zcnsc/mint_test.go (hunk 4 above)
Comment (areebusmani) [reply code/go/0chain.net/smartcontract/zcnsc/mint_test.go:175]: Fixed

Comment (areebusmani) [reply code/go/0chain.net/smartcontract/zcnsc/mint_test.go:175]: The error will still be "signatures not found" because since numAuth is zero so signatures will be sliced to an empty array.

https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/zcnsc/mint.go#L62

File: code/go/0chain.net/smartcontract/zcnsc/mint.go (hunk 2 above)
Comment (areebusmani) [reply code/go/0chain.net/smartcontract/zcnsc/mint.go:51]: Actually, the panic does not happen in the unit test case that I have added in this PR. 

Since number of authorizers is 0 so the signatures are sliced to an empty array. 
https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/zcnsc/mint.go#L62

Then, while trying to verify the signatures the method is returning an error saying that there are no signatures.

https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/zcnsc/mint.go#L99
https://github.com/0chain/0chain/blob/31c960fcf9d2ebc1e160e8810b28927132f2ce90/code/go/0chain.net/smartcontract/zcnsc/models.go#L154

Hence, we don't get to the point where we divide the share to all authorizers.

Previously, we might have been getting panic because of the missing return statement here.
https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/zcnsc/mint.go#L102

File: code/go/0chain.net/smartcontract/zcnsc/context_helpers_test.go [hunk 5]
```diff
@@ -239,6 +239,192 @@ func MakeMockStateContext() *mockStateContext {
 	return ctx
 }
 
+func MakeMockStateContextWithNoAutorizers() *mockStateContext {
```
Comment (areebusmaniblr) [inline code/go/0chain.net/smartcontract/zcnsc/context_helpers_test.go:242]: @peterlimg Since golang neither supports optional parameters nor method overloading so I was finding it difficult to add modify MakeMockStateContext function to conditional add or not add the authorizers. So I copied the method to a new method MakeMockStateContextWithNoAutorizers and removed the part responsible for adding autorizers.

Can you suggest a better way of achieving the same thing?

File: code/go/0chain.net/smartcontract/zcnsc/mint.go (hunk 2 above)
Comment (peterlimg) [reply code/go/0chain.net/smartcontract/zcnsc/mint.go:51]: @areebusmaniblr yes, that should work to fix the panic (thought this branch would never panic as it does not have the latest code from staging), but we'd better check the 0 number of authorizer explicitly instead of letting the empty sliced signatures' verification code to do the 0 checking, in any chance if the `verifySignature` does not do the empty checking, we will panic here https://github.com/0chain/0chain/blob/59c0b3a1481d8e3c776c612987f494270d256e61/code/go/0chain.net/smartcontract/zcnsc/mint.go#L121. Anyway, I'll do a quick commit and merge this PR.

**Review by peterlimg — APPROVED** [review]
LGTM
